<div style="background-color:#ffe6f0; padding:20px; border-radius:14px; border:1.5px solid #ffb3d1; color:#1a1a1a;">

<h3 style="margin-top: 0; color: #c2185b;"> Fase I: Formalismo matemático y diferenciación numérica</h3>

Sea una función $f(x)$ aproximada por interpolación de Lagrange en nodos $x_j = x_0 + jh$, $j=0,\dots,n$:

$$
P_n(x)=\sum_{j=0}^{n} l_j(x)f_j, \quad f_j=f(x_j),
$$

donde

$$
l_j(x)=\prod_{\substack{m=0 \\ m\neq j}}^{n}\frac{x-x_m}{x_j-x_m}.
$$

La derivada numérica general se obtiene derivando el polinomio:

$$
f'(x)\approx P_n'(x)=\sum_{j=0}^{n} l_j'(x)f_j.
$$

Evaluando en $x=x_0+sh$, se escribe como:

$$
f'(x_0+sh)\approx \frac{1}{h}\sum_{j=0}^{n} w_j(s)f_j.
$$

Para $n=2$ (fórmula de 3 puntos), la primera derivada es:

$$
f'(x_0+sh)\approx
\frac{(-3+2s)f_0+4(1-s)f_1+(-1+2s)f_2}{2h}.
$$

Caso especial $s=1$:

$$
f'(x_1)\approx \frac{f_2-f_0}{2h},
$$

que corresponde a la diferencia central estándar.

La segunda derivada de 3 puntos es:

$$
f''(x_1)\approx \frac{f_0-2f_1+f_2}{h^2}.
$$

</div>

In [2]:
import numpy as np

def first_derivative_three_point(f0, f1, f2, h, s=1.0):
    return ((-3 + 2*s)*f0 + 4*(1-s)*f1 + (-1 + 2*s)*f2)/(2*h)

def second_derivative_three_point(f0, f1, f2, h):
    return (f0 - 2*f1 + f2)/(h**2)


In [3]:
def f(x): return x**3
def fp_exact(x): return 3*x**2
def fpp_exact(x): return 6*x

x0 = 1.0
h = 1e-3
s = 1.0

x1 = x0 + h
x2 = x0 + 2*h

f0, f1, f2 = f(x0), f(x1), f(x2)

fp_num = first_derivative_three_point(f0, f1, f2, h, s=s)  # en x1 si s=1
fpp_num = second_derivative_three_point(f0, f1, f2, h)     # en x1

print("f'(x1) num:", fp_num, " exact:", fp_exact(x1), " error:", abs(fp_num - fp_exact(x1)))
print("f''(x1) num:", fpp_num, " exact:", fpp_exact(x1), " error:", abs(fpp_num - fpp_exact(x1)))


f'(x1) num: 3.006003999999951  exact: 3.006002999999999  error: 9.999999521781433e-07
f''(x1) num: 6.006000000446932  exact: 6.005999999999999  error: 4.469322689715227e-10


In [6]:
import pandas as pd
from IPython.display import display

data = {
    "Magnitud": ["f'(x1)", "f''(x1)"],
    "Valor numerico": [fp_num, fpp_num],
    "Valor exacto": [fp_exact(x1), fpp_exact(x1)],
    "Error absoluto": [abs(fp_num - fp_exact(x1)), abs(fpp_num - fpp_exact(x1))]
}

df_fase1 = pd.DataFrame(data)

styled_fase1 = (
    df_fase1.style
    .format({
        "Valor numerico": "{:.12f}",
        "Valor exacto": "{:.12f}",
        "Error absoluto": "{:.3e}",
    })
    .background_gradient(subset=["Error absoluto"], cmap="RdPu_r")
    .set_properties(**{"text-align": "center"})
    .set_table_styles([
        {"selector": "table", "props": [("border-collapse", "collapse"), ("border", "1px solid #b0b0b0")]},
        {"selector": "th", "props": [("border", "1px solid #b0b0b0"), ("padding", "6px"), ("background-color", "#f8d7e8")]},
        {"selector": "td", "props": [("border", "1px solid #b0b0b0"), ("padding", "6px")]},
        {"selector": "caption", "props": [("caption-side", "top"), ("font-weight", "bold"), ("font-size", "14px")]}
    ])
    .set_caption("Validacion numerica - Fase I (Derivadas de 3 puntos)")
)

display(styled_fase1)



,Magnitud,Valor numerico,Valor exacto,Error absoluto
0,f'(x1),3.006004000000,3.006003000000,1.000e-06
1,f''(x1),6.006000000447,6.006000000000,4.469e-10


<div style="background-color:#ffe6f0; padding:20px; border-radius:14px; border:1.5px solid #ffb3d1; color:#1a1a1a;">

<h3 style="margin-top: 0; color: #c2185b;">💖 Conclusión de validación (Fase I)</h3>

Se validaron las fórmulas de diferenciación numérica de 3 puntos con la función prueba $f(x)=x^3$, usando $x_0=1$, $h=10^{-3}$ y $s=1$.

<br>

<b>Resultados:</b>

<ul>
<li>Primera derivada: error absoluto $\approx 10^{-6}$.</li>
<li>Segunda derivada: error absoluto $\approx 4.47\times10^{-10}$.</li>
</ul>

Los errores son pequeños y consistentes con el comportamiento esperado de los esquemas de diferencias finitas, por lo tanto la implementación de Fase I queda verificada.

</div>